src/publisher.js

```diff
@@ -0,0 +1,436 @@
+const assert = require('assert');
+const libUrls = require('taskcluster-lib-urls');
+const debug = require('debug')('taskcluster-lib-pulse.publisher');
+const EventEmitter = require('events');
+
+class Exchanges {
+  constructor(options) {
+    assert(options.serviceName, 'serviceName is required');
+    assert(options.projectName, 'projectName is required');
+    assert(options.version, 'version is required');
+    assert(options.title, 'title is required');
+    assert(options.description, 'description is required');
+
+    Object.assign(this, options);
+    this.entries = [];
+
+    this.exchangePrefix = `exchange/${this.projectName}/${this.version}/`;
+  }
+
+  declare(entryOptions) {
+    const entry = new Entry({exchanges: this, ...entryOptions});
+    assert(!this.entries.some(e => e.name == entry.name),
+      `entry with name ${entry.name} already declared`);
+    assert(!this.entries.some(e => e.exchange == entry.exchange),
+      `entry with exchange ${entry.exchange} already declared`);
+    this.entries.push(entry);
+  }
+
+  reference() {
+    return {
+      version: 0,
+      $schema: 'http://schemas.taskcluster.net/base/v1/exchanges-reference.json#',
+      serviceName: this.serviceName,
+      title: this.title,
+      description: this.description,
+      exchangePrefix: this.exchangePrefix,
+      entries: this.entries.map(e => e.reference()),
+    };
+  }
+
+  async publisher({rootUrl, schemaset, client, sendDeadline, publish, aws}) {
+    let publisher;
+    if (client.isFakeClient) {
+      publisher = new FakePulsePublisher({rootUrl, schemaset, exchanges: this});
+    } else {
+      publisher = new PulsePublisher({rootUrl, schemaset, client, sendDeadline, exchanges: this});
+    }
+    if (publish) {
+      assert.equal(rootUrl, 'https://taskclutser.net',
+        'only taskcluster.net publishes references to S3');
+      assert(aws, 'aws is required to publish references to S3');
+      await publisher.publishReference(aws, this.reference());
+    }
+    await publisher._start();
+    return publisher;
+  }
+}
+
+exports.Exchanges = Exchanges;
+
+class Entry {
+  constructor({exchanges, ...options}) {
+    assert(options.exchange, 'exchange is required');
+    assert(options.name, 'name is required');
+    assert(options.title, 'title is required');
+    assert(options.description, 'description is required');
+    assert(options.schema, 'schema is required');
+    assert(options.routingKey, 'routingKey is required');
+    assert(options.messageBuilder, 'messageBuilder is required');
+    assert(options.routingKeyBuilder, 'routingKeyBuilder is required');
+    assert(options.CCBuilder, 'CCBuilder is required');
+
+    Object.assign(this, options);
+    this.exchanges = exchanges;
+
+    // perform this transformation once, to a URI relative to the service's schemas
+    this.schema = `${exchanges.version}/${this.schema.replace(/\.ya?ml$/, '.json#')}`;
```

Comment:
I feel like we do this in a lot of places. Maybe we should add it to tc-lib-urls?

Reply:
Yeah, we had it there, but there *are* things that don't have a version (such as common schemas), so it would mean two different schema functions..

I think this could get better fixed in some future `taskcluster-lib-references` that handles references and schemas, knows how to do this filename-to-schema-name translation, knows how to generate schema URLs, etc.  :handwave: <-- why isn't that an emoji?!?!